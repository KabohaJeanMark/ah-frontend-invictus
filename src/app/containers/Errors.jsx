import React, { Component } from 'react';

export default class Errors extends Component {
  constructor (props) {
    super(props);
    this.state = { errors: [] };
  }
  componentWillReceiveProps (nextProps) {
    if (nextProps.errors) {
      this.setState({
        errors: this.mapErrorsToList(nextProps)
      });
    }
  }
   mapErrorsToList = props => {
     let items = [];
     if (Array.isArray(props.errors)) {
       for (let i = 0; i < props.errors.length; i++) {
         items.push(props.errors[i]);
       }
     } else {
       Object.keys(props.errors).forEach(key => {
         for (let i = 0; i < props.errors[key].length; i++) {
           items.push(key + ': ' + props.errors[key][i]);
         }
       });
     }
     return items;
   }
   render () {
     function Repeat (props) {
       let items = [];
       for (let i = 0; i < props.errors.length; i++) {
         items.push(props.children(i, props.errors[i]));
       }
       return <div className="danger" ><ul>{items}</ul></div>;
     }
     return (
       this.state.errors.length ? (
         <Repeat errors={this.state.errors}>
           {(index, message) => <li key={index}>{message}</li>}
         </Repeat>
       ) : (null)
     );
   }
}
