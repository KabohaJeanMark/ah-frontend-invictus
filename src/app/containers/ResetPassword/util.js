export const myComponentWillReceiveProps = nextProps => {
  let stateObj;
  if (nextProps.message) {
    stateObj = { message: nextProps.message };
  }
  if (nextProps.errors) {
    stateObj = { ...stateObj, errors: nextProps.errors };
  } else {
    stateObj = { ...stateObj, email: "" };
  }
  return stateObj;
};
