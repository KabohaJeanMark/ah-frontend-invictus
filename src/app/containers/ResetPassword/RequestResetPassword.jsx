import React, { Component } from 'react';
import { resetPasswordRequestAction } from '../../../redux/actions/resetPasswordAction';
import PasswordResetForm from '../../components/PasswordResetForm';
import { connect } from 'react-redux';
import { APP_URLS } from '../../../utilities';
import { myComponentWillReceiveProps } from './util';

export class ResetPassword extends Component {
  constructor (props) {
    super(props);
    this.state = {
      email: "",
      errors: [],
      message: "",
      form: "email",
      isDisplayMessage: false
    };
  }
  componentWillReceiveProps (ReceivedProps) {
    this.setState(myComponentWillReceiveProps(ReceivedProps));
  }
    onChange = event => {
      this.setState({ [event.target.name]: event.target.value });
    }
    onSubmit = e => {
      e.preventDefault();
      const post = {
        email: this.state.email,
        url: APP_URLS.FRONT_END + '/resetpassword/'
      };
      this.setState({
        isDisplayMessage: true,
        message: "Please wait .."
      });
      this.props.resetPasswordRequestAction(post);
    }
    render () {
      const props = {
        onSubmit: this.onSubmit,
        onChange: this.onChange,
        state: this.state
      };
      return (
        this.state.message === "Check your email-address for a reset-password link." ?
          this.state.message :
          <PasswordResetForm props={props} />
      );
    }
}
export const mapStateToProps = state => ({
  message: state.RequestResetPassword.message,
  errors: state.RequestResetPassword.errors
});
export default connect(mapStateToProps, { resetPasswordRequestAction })(ResetPassword);
