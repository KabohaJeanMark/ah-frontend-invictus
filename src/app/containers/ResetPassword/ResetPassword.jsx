import React, { Component } from 'react';
import { resetPasswordAction } from '../../../redux/actions/resetPasswordAction';
import PasswordResetForm from '../../components/PasswordResetForm';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { myComponentWillReceiveProps } from './util';
import { Redirect } from 'react-router-dom';

export class ResetPassword extends Component {
  constructor (props) {
    super(props);
    this.state = {
      password: "",
      password2: "",
      errors: [],
      token: "",
      message: "",
      form: "password",
      isDisplayMessage: false
    };
  }
  componentDidMount () {
    const { match: { params } } = this.props;
    this.setState({ token: params.token });
  }
  componentWillReceiveProps (nextProps) {
    this.setState(myComponentWillReceiveProps(nextProps));
  }
    onChange = event => {
      this.setState({ [event.target.name]: event.target.value });
    }
    onSubmit = e => {
      e.preventDefault();
      const post = {
        token: this.state.token,
        password: this.state.password,
        confirm_password: this.state.password2
      };
      this.setState({
        isDisplayMessage: true,
        message: "Please wait .."
      });
      this.props.resetPasswordAction(post);
    }
    render () {
      return (
        this.state.message === "Password reset was successfull." ?
          <Redirect to="/login" /> :
          <PasswordResetForm
            props={{ onSubmit: this.onSubmit,
              onChange: this.onChange,
              state: this.state }} />
      );
    }
}
ResetPassword.propTypes = {
  message: PropTypes.string,
  errors: PropTypes.array,
  match: PropTypes.object,
  resetPasswordAction: PropTypes.func
};
const mapStateToProps = state => ({
  message: state.RequestResetPassword.message,
  errors: state.RequestResetPassword.errors
});
export default connect(mapStateToProps, { resetPasswordAction })(ResetPassword);
