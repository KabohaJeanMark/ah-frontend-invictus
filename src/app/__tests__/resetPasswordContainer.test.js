import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { ResetPassword } from '../containers/ResetPassword/ResetPassword';
import { ResetPassword as RequestResetPassword } from '../containers/ResetPassword/RequestResetPassword';
import Errors from '../containers/Errors';

describe('<ResetPassword />', () => {
  const props = {
    message: "here",
    errors: ["ada"],
    resetPasswordAction: a => a,
    match: { params: { token: "token" } }
  };
  const wrapper = shallow(<ResetPassword {...props} />);

  it('onclick should update state', () => {
    const event = {
      target: {
        name: "password",
        value: "password"
      }
    };
    wrapper.instance().onChange(event);
    expect(wrapper.instance().state.password).toEqual('password');
  });

  it('onsubmit should update state', () => {
    const event = {
      preventDefault: jest.fn()
    };
    wrapper.instance().onSubmit(event);
    expect(wrapper.instance().state.isDisplayMessage).toBeTruthy();
    expect(wrapper.instance().state.message).toBe("Please wait ..");
  });

  it('componentWillReceiveProps should update state', () => {
    wrapper.instance().componentWillReceiveProps({});
    wrapper.instance().componentWillReceiveProps(props);
    expect(wrapper.instance().state.errors).toEqual(["ada"]);
    expect(wrapper.instance().state.message).toBe("here");
  });
});

describe('<RequestResetPassword />', () => {
  const props = {
    message: "here",
    errors: ["ada"],
    resetPasswordRequestAction: a => a
  };
  const wrapper = shallow(<RequestResetPassword {...props} />);

  it('onclick should update state', () => {
    const event = {
      target: {
        name: "email",
        value: "email"
      }
    };
    wrapper.instance().onChange(event);
    expect(wrapper.instance().state.email).toEqual('email');
  });

  it('onsubmit should update state', () => {
    const event = {
      preventDefault: jest.fn()
    };
    wrapper.instance().onSubmit(event);
    expect(wrapper.instance().state.isDisplayMessage).toBeTruthy();
    expect(wrapper.instance().state.message).toBe("Please wait ..");
  });

  it('componentWillReceiveProps should update state', () => {
    wrapper.instance().componentWillReceiveProps({});
    wrapper.instance().componentWillReceiveProps(props);
    expect(wrapper.instance().state.errors).toEqual(["ada"]);
    expect(wrapper.instance().state.message).toBe("here");
  });
});

describe('<Errors />', () => {
  let wrapper;
  let state = {
    errors: ['Error message']
  };
  wrapper = mount(<Errors props={state} />);

  it('matches snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('componentWillReceiveProps should update state', () => {
    wrapper.instance().componentWillReceiveProps({});
    wrapper.instance().componentWillReceiveProps({ errors: { error: ['Error message'] } });
    wrapper.instance().componentWillReceiveProps(state);
    expect(wrapper.instance().state.errors).toEqual(["Error message"]);
  });
});
