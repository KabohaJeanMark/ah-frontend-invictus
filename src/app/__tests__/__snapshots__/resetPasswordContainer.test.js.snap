// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Errors /> matches snapshot 1`] = `
<Errors
  props={
    Object {
      "errors": Array [
        "Error message",
      ],
    }
  }
/>
`;
